bundle/src/main/java/com/adobe/acs/commons/remoteassets/impl/RemoteAssetsNodeSyncImpl.java

```diff
@@ -0,0 +1,609 @@
+        this.resourceResolver = RemoteAssets.logIn(this.resourceResolverFactory);
+        this.session = this.resourceResolver.adaptTo(Session.class);
+        if (StringUtils.isNotBlank(remoteAssetsConfig.getEventUserData())) {
+            this.session.getWorkspace().getObservationManager().setUserData(this.remoteAssetsConfig.getEventUserData());
+        }
+        this.valueFactory = this.session.getValueFactory();
+    }
+
+    /**
+     * Method to run on deactivation.
+     */
+    @Deactivate
+    protected void deactivate() {
+        if (this.session != null) {
+            try {
+                this.session.logout();
+            } catch (Exception e) {
+                LOG.warn("Failed session.logout()", e);
+            }
+        }
+
+        if (this.resourceResolver != null) {
+            try {
+                this.resourceResolver.close();
+            } catch (Exception e) {
+                LOG.warn("Failed resourceResolver.close()", e);
+            }
+        }
+    }
+
+    /**
+     * @see RemoteAssetsNodeSync#syncAssetNodes()
+     */
+    @Override
+    public void syncAssetNodes() {
+        List<String> syncPaths = new ArrayList<>();
+        syncPaths.addAll(this.remoteAssetsConfig.getTagSyncPaths());
+        syncPaths.addAll(this.remoteAssetsConfig.getDamSyncPaths());
+
+        try {
+            for (String syncPath : syncPaths) {
+                this.session.refresh(true);
+                JSONObject topLevelJson = getJsonFromUri(syncPath);
+                Node topLevelSyncNode = getOrCreateNode(syncPath, (String) topLevelJson.get(JcrConstants.JCR_PRIMARYTYPE));
+                createOrUpdateNodes(topLevelJson, topLevelSyncNode);
+                this.session.save();
+            }
+        } catch (IOException e) {
+            LOG.error("IO Exception {}", e);
+        } catch (JSONException e) {
+            LOG.error("Json Exception {}", e);
+        } catch (RepositoryException e) {
+            LOG.error("Repository Exception {}", e);
+        }
+    }
+
+    /**
+     * Grab JSON from the provided path and sync to the JCR.
+     * @param nextPath String
+     * @param primaryType String
+     * @return Node
+     */
+    private Node getOrCreateNode(final String nextPath, final String primaryType) {
+        Node node = null;
+
+        try {
+            node = JcrUtils.getNodeIfExists(nextPath, this.session);
+            if (node == null) {
+                node = JcrUtil.createPath(nextPath, primaryType, this.session);
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("New node '{}' created.", node.getPath());
+                }
+            } else {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Node '{}' retrieved from JCR.", node.getPath());
+                }
+            }
+        } catch (RepositoryException e) {
+            LOG.error("Repository Exception. Unable to get or create node '{}'", nextPath, e);
```

Comment:
We might want to re-throw this exception, because nothing that calls this function is handling the case if node is `null` and honestly if this actually does happen we probably have a big enough problem that nothing is going to work.